### PR TITLE
Fix start_time for cancelled jobs and sort scripts/jobs

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
The list outputs are now sorted according to:

* `scripts`: By the `id`/`name` as this can be set by the end user, and
* `jobs`: By the `created_at` time

---

The `start_time` time is no longer set for jobs which are cancelled whilst in a pending state. This is handled by the `slurm/monitor.sh` script to prevent issues with polling intervals.

Jobs which were not allocated resources are "assumed" not to have started. This condition is detected in two ways:
* With `scontrol` when the `NodeList` is empty (technically `(null)`), and
* When `AllocTRES` field is empty in `sacct`.

### CODE REVIEW

File: libexec/slurm/monitor.sh [hunk 1]
```diff
@@ -49,21 +49,40 @@ read -r -d '' template <<'TEMPLATE' || true
...
 
 # Fetch the state of the job
-control=$(scontrol show job "$1" --oneline | head -n 1 | tr ' ' '\n' 2>&1)
+raw_control=$(scontrol show job "$1" --oneline 2>&1)
 exit_status="$?"
+control=$(echo "$raw_control" | head -n 1 | tr ' ' '\n')
+cat <<EOF
+scontrol:
+$control
+EOF
```
Comment (benarmston) [inline libexec/slurm/monitor.sh:58]: Is this debugging code that should be removed?

File: libexec/slurm/monitor.sh [hunk 2]
```diff
@@ -49,21 +49,40 @@ read -r -d '' template <<'TEMPLATE' || true
...
   # Fallback to sacct if scontrol does not recognise the ID
-  acct=$(sacct --noheader --parsable --jobs "$1" --format State,Reason,START,END  | head -n1)
+  raw_acct=$(sacct --noheader --parsable --jobs "$1" --format State,Reason,START,END,AllocTRES)
   exit_status="$?"
+  acct=$(echo "$raw_acct" | head -n1)
+  cat <<EOF
+
+sacct:
+$raw_acct
+EOF
```
Comment (benarmston) [inline libexec/slurm/monitor.sh:85]: Is this debugging code that should be removed?

File: libexec/slurm/monitor.sh [hunk 3]
```diff
@@ -73,11 +92,18 @@ elif [[ "$control" == "slurm_load_jobs error: Invalid job id specified" ]]; then
...
 
   # Extract the output from sacct
-  elif [[ "$exit_status" -eq 0 ]]; then
-    state=$(echo "$acct" | cut -d'|' -f1)
-    reason=$(echo "$acct" | cut -d'|' -f2)
+elif [[ "$exit_status" -eq 0 ]]; then
+  # sacct sometimes tacks info onto the end of the state
+  state=$(echo "$acct" | cut -d'|' -f1 | cut -d' ' -f1)
+  reason=$(echo "$acct" | cut -d'|' -f2)
+  if [ "$(echo "$acct" | cut -d'|' -f5)" ]; then
```
Comment (benarmston) [inline libexec/slurm/monitor.sh:99]: It would be better to make the comparison to the empty string explicit here.  `if [ "$( ... )" != "" ]; then`.

**Review by benarmston — CHANGES_REQUESTED** [review]
Could you take a look at the comments I've added @WilliamMcCumstie .

File: libexec/slurm/monitor.sh (hunk 2 above)
Comment (WilliamMcCumstie) [reply libexec/slurm/monitor.sh:85]: It is debugging code, but I decided to leave it in so it becomes part of the log file when `log_level` is set to `DEBUG`.

The JSON payload is limited to the last line, so it doesn't make a difference. However I can redirect it to STDERR as an alternative.

Comment (benarmston) [reply libexec/slurm/monitor.sh:85]: Please redirect it to STDERR.

File: libexec/slurm/monitor.sh (hunk 3 above)
Comment (benarmston) [reply libexec/slurm/monitor.sh:99]: ping.

File: libexec/slurm/monitor.sh (hunk 2 above)
Comment (benarmston) [reply libexec/slurm/monitor.sh:85]: ping

File: libexec/slurm/monitor.sh (hunk 1 above)
Comment (benarmston) [reply libexec/slurm/monitor.sh:58]: ping

File: libexec/slurm/monitor.sh (hunk 2 above)
Comment (WilliamMcCumstie) [reply libexec/slurm/monitor.sh]: https://github.com/openflighthpc/flight-job/pull/14/commits/8dbac91b8f831737f4e38e773d163dac7a27bbc5

File: libexec/slurm/monitor.sh (hunk 1 above)
Comment (WilliamMcCumstie) [reply libexec/slurm/monitor.sh]: https://github.com/openflighthpc/flight-job/pull/14/commits/8dbac91b8f831737f4e38e773d163dac7a27bbc5

File: libexec/slurm/monitor.sh (hunk 2 above)
Comment (WilliamMcCumstie) [reply libexec/slurm/monitor.sh]: :facepalm:

Comment (WilliamMcCumstie) [reply libexec/slurm/monitor.sh]: https://github.com/openflighthpc/flight-job/pull/14/commits/bd68dd00f25388794df831dc5bc8e97ca9087545

File: libexec/slurm/monitor.sh (hunk 1 above)
Comment (WilliamMcCumstie) [reply libexec/slurm/monitor.sh]: https://github.com/openflighthpc/flight-job/pull/14/commits/bd68dd00f25388794df831dc5bc8e97ca9087545

File: libexec/slurm/monitor.sh (hunk 3 above)
Comment (WilliamMcCumstie) [reply libexec/slurm/monitor.sh]: The idiomatic way to do this in `bash` is to use the `-n` flag:

https://github.com/openflighthpc/flight-job/pull/14/commits/34c9a3cffe21b9d5dd7790319e979f5f9a271433

Comment (benarmston) [reply libexec/slurm/monitor.sh]: `-n` is good too.  As long as we're not relying on the reader to recall that `[ "" ]` is false.

**Comment by benarmston** [comment]
LGTM now.

**Pull request merged**